Src/FluentAssertions.Json/JTokenDifferentiator.cs

```diff
@@ -26,34 +26,84 @@ public static Difference FindFirstDifference(JToken actual, JToken expected)
                 return new Difference(DifferenceKind.ExpectedIsNull, path);
             }
             
-            return FindFirstDifference(actual, expected, path);
+            return FindFirstDifference(actual, expected, path, ignoreExtraProperties);
         }
 
-        private static Difference FindFirstDifference(JToken actual, JToken expected, JPath path)
+        private static Difference FindFirstDifference(JToken actual, JToken expected, JPath path, bool ignoreExtraProperties)
         {
             switch (actual)
             {
                 case JArray actualArray:
-                    return FindJArrayDifference(actualArray, expected, path);
-                case JObject actualObbject:
-                    return FindJObjectDifference(actualObbject, expected, path);
+                    return FindJArrayDifference(actualArray, expected, path, ignoreExtraProperties);
+                case JObject actualObject:
+                    return FindJObjectDifference(actualObject, expected, path, ignoreExtraProperties);
                 case JProperty actualProperty:
-                    return FindJPropertyDifference(actualProperty, expected, path);
+                    return FindJPropertyDifference(actualProperty, expected, path, ignoreExtraProperties);
                 case JValue actualValue:
                     return FindValueDifference(actualValue, expected, path);
                 default: 
                     throw new NotSupportedException();
             }
         }
 
-        private static Difference FindJArrayDifference(JArray actualArray, JToken expected, JPath path)
+        private static Difference FindJArrayDifference(JArray actualArray, JToken expected, JPath path,
+            bool ignoreExtraProperties)
         {
             if (!(expected is JArray expectedArray))
             {
                 return new Difference(DifferenceKind.OtherType, path, Describe(actualArray.Type), Describe(expected.Type));
             }
-            
-            return CompareItems(actualArray, expectedArray, path);
+
+            if (ignoreExtraProperties)
+            {
+                return CompareExpectedItems(actualArray, expectedArray, path);
+            }
+            else
+            {
+                return CompareItems(actualArray, expectedArray, path);
+            }
+        }
+
+        private static Difference CompareExpectedItems(JArray actual, JArray expected, JPath path)
+        {
+            JEnumerable<JToken> actualChildren = actual.Children();
+            JEnumerable<JToken> expectedChildren = expected.Children();
+
+            int matchingIndex = 0;
+            for (int expectedIndex = 0; expectedIndex < expectedChildren.Count(); expectedIndex++)
+            {
+                var expectedChild = expectedChildren.ElementAt(expectedIndex);
+                bool match = false;
+                for (int actualIndex = matchingIndex; actualIndex < actualChildren.Count(); actualIndex++)
```

Comment:
As `actualChildren` is iterated in an inner loop, we might want to materialize the enumerable, to avoid re-calculations.